[CAS] gmodule support for caching build

cachemeifyoucan opened the pull request:
Prototype for gmodule support by switching `splitDwarf` references to CASIDs of the module/PCH

Also teach dsymutil to support reading from CAS as a prototype for lldb support.

adrian-prantl commented on lldb/scripts/clang-explicit-module-build.py:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
```
This would potentially fit better into llvm-project/lldb/packages/Python/lldbsuite/

cachemeifyoucan replied:
This is a compiler launcher type of script and requires a just built clang (because it needs clang-scan-deps next to it). I don't what is a good way to hack into lldbsuite or driven from makefile. Let me know if you have any good idea.

cachemeifyoucan commented:
@adrian-prantl @benlangmuir Ping for review

benlangmuir commented on clang/include/clang/Basic/Module.h:
```diff
@@ -274,6 +274,9 @@ class alignas(8) Module {
   std::optional<std::string> ModuleCacheKey;
 
+  /// The \c CASID for the loaded module, otherwise empty.
+  std::string CASID;
```
Nit: we use `std::optional<std::string>` above for the key; we should be consistent about whether to use optional or just empty string.

benlangmuir commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -6150,7 +6168,7 @@ bool ASTReader::readASTFileControlBlock(
       StringRef CacheKey = ReadStringBlob(Record, Idx, Blob);
       if (!CacheKey.empty())
-        Listener.readModuleCacheKey(ModuleName, *Filename, CacheKey);
+        Listener.readModuleCASID(ModuleName, *Filename, CacheKey);
```
How do we know here that this is not a key?

cachemeifyoucan replied:
This code path is never used tor read scanning module. I will rename the variable. Let me think if there is a way to guard it against using it to read scanning module (other than this callback will fail).

benlangmuir replied:
Would it make sense to store a bit in the pcm for whether it's a key or just an id? Even if it's not strictly necessary it might make it less likely we screw it up in the future

cachemeifyoucan commented:
@swift-ci please test llvm

cachemeifyoucan commented:
@swift-ci please test llvm

cachemeifyoucan committed:
[clang][gmodules] Use CASID to encode splitDwarfFilename

Teach clang to encode CASID as splitDwarfFilename for gmodule when
clang caching is enabled. This allows the outputs from compiler do not
contain paths to the clang module files, thus allows distributed caching
without the need of a unified clang module cache directory path.

cachemeifyoucan committed:
[dsymutil] Teach dsymutil about CAS and loading clang modules from CAS

Teach dsymutil how to use CAS and how to load clang modules from CAS
when building dSYM when gmodule is used.

cachemeifyoucan committed:
[lldb] Teach LLDB to load gmodules from CAS

Teach lldb to load clang modules when gmodule + clang caching is used.

cachemeifyoucan commented:
@swift-ci please test llvm

benlangmuir commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3401,6 +3401,11 @@ llvm::DIModule *CGDebugInfo::getOrCreateModuleRef(ASTSourceDescriptor Mod,
     }
     llvm::sys::path::append(PCM, Mod.getASTFile());
+
+    // FIXME: Prefer CASID if exists.
```
I'm not clear what this FIXME means

cachemeifyoucan replied:
I don't know where there is a FIXME here. Removed.

benlangmuir commented on llvm/tools/dsymutil/Options.td:
```diff
@@ -218,3 +218,8 @@ def dsym_search_path: Separate<["-", "--"], "D">,
   HelpText<"Specify a directory that contain dSYM files to search for.">,
   Group<grp_general>;
+
+def cas : Separate<["-", "--"], "cas">,
```
Should we provide the additional options needed to configure plugin path and plugin options?

benlangmuir reviewed (approved):
LGTM for the clang changes. Someone else should review the LLDB and dsymutil code.

adrian-prantl commented on clang/test/ClangScanDeps/gmodules.c:
```diff
@@ -0,0 +1,228 @@
+/// Check debug info is correct.
+// RUN: %clang %t/tu.o -o %t/a.out
+// RUN: dsymutil -cas %t/cas %t/a.out -o %t/a.dSYM 2>&1 | FileCheck %s --check-prefix=WARN --allow-empty
+// RUN: dsymutil %t/a.out -o %t/a2.dSYM 2>&1 | FileCheck %s --check-prefix=WARN --allow-empty
```
Does this being `dsymutil` and not `%dsymutil` mean we might pick up the system one? Since there is also a dsymutil aspect to this, could that be problematic in CI?

cachemeifyoucan replied:
Tools are always in the path.

NOTE: this is not the only test that calls dsymutil in clang tests and dsymutil wasn't even a build dependency for tests (added in this PR).

adrian-prantl commented on llvm/tools/dsymutil/Options.td:
```diff
@@ -218,3 +218,16 @@ def dsym_search_path: Separate<["-", "--"], "D">,
+
+def cas : Separate<["-", "--"], "cas">,
+  MetaVarName<"<path>">,
+  HelpText<"Specify CAS directory to load CAS references from.">,
```
Does this imply that there can only be one CAS per build?
Is that realistic? What if people use static archives and combine objects from different projects?

cachemeifyoucan replied:
Yes, it has to be one CAS per build. It is a requirement for module imports and there is no way you can configure otherwise.

If you have multiple builds and uses some products from other build, the second CAS has to ingest everything from previous builds.

adrian-prantl commented on llvm/tools/dsymutil/BinaryHolder.cpp:
```diff
@@ -231,6 +264,22 @@ BinaryHolder::ArchiveEntry::getObjectEntry(StringRef Filename,
+  WithColor::note() << "create CAS using configuration: " << Config->first
+                    << "'\n";
+  return DB->first;
+}
```
It would be good to find a way to mock this up in a dsymutil-only test that doesn't depend on clang.

cachemeifyoucan replied:
Let me know if you have good ways to do it. It is very hard to do it since it needs to construct both object files (which can be committed) but also module files and CAS contents.

adrian-prantl replied:
Usually we check in binary files for dsymutil tests. We only need to resolve a single URL to test it. Is there maybe room for a `null` CAS plugin where the llvmcas:// URL is a filename directly?

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -282,6 +286,32 @@ bool ModuleListProperties::GetSwiftEnableASTContext() const {
+// START CAS
+FileSpec ModuleListProperties::GetCASOnDiskPath() const {
+  const uint32_t idx = ePropertyCASOnDiskPath;
+  return GetPropertyAtIndexAs<FileSpec>(idx, {});
```
How would these settings be communicated to LLDB in the typical use-case?
What if a debug session contains multiple dylibs built against different CAS storage? (I.e., in each project's DerivedData/)

cachemeifyoucan replied:
See `ConfigureCASStorage()`. 

This option provides a global overwrite. If no global setting available, it will search for a CAS configuration from the path of the debugging binary upwards to find a CAS configuration file.

cachemeifyoucan commented:
Create PR on stable branch (so test is available) if you like to review over there: https://github.com/swiftlang/llvm-project/pull/11714

adrian-prantl commented on llvm/tools/dsymutil/BinaryHolder.cpp:
```diff
@@ -231,6 +264,22 @@ BinaryHolder::ArchiveEntry::getObjectEntry(StringRef Filename,
+static std::shared_ptr<cas::ObjectStore> searchAndCreateCAS(StringRef Path) {
+  auto Config = cas::CASConfiguration::createFromSearchConfigFile(Path);
+  if (!Config)
+    return nullptr;
```
Can we make sure this returns an llvm::Error explaining that the config file could not be found?

adrian-prantl commented on llvm/tools/dsymutil/BinaryHolder.cpp:
```diff
@@ -231,6 +264,22 @@ BinaryHolder::ArchiveEntry::getObjectEntry(StringRef Filename,
+  auto Config = cas::CASConfiguration::createFromSearchConfigFile(Path);
+  if (!Config)
+    return nullptr;
+
```
Similarly, there should be an error displayed if the cas format is too new/old for dsymutil to understand.

adrian-prantl commented on llvm/tools/dsymutil/dsymutil.cpp:
```diff
@@ -666,12 +681,24 @@ int dsymutil_main(int argc, char **argv, const llvm::ToolContext &) {
+    if (!DB) {
+      WithColor::error() << "failed to open CAS: " << toString(DB.takeError())
+                         << "\n";
+      return EXIT_FAILURE;
```
Can we add tests that check the error handling when:
- no config file is found
- the cas in the config file is corrupted/incompatible/...
- the url cannot be resolved in the cas

adrian-prantl commented on llvm/tools/dsymutil/Options.td:
```diff
@@ -218,3 +218,16 @@ def dsym_search_path: Separate<["-", "--"], "D">,
+  MetaVarName<"<path>">,
+  HelpText<"Path to CAS plugin library if used">,
+  Group<grp_general>;
+def cas_plugin_option : Separate<["-", "--"], "cas-plugin-option">,
```
Can you also update llvm/docs/CommandGuide/dsymutil.rst?

adrian-prantl commented on lldb/include/lldb/Core/ModuleList.h:
```diff
@@ -522,6 +528,11 @@ class ModuleList {
                   llvm::SmallVectorImpl<lldb::ModuleSP> *old_modules,
                   bool *did_create_ptr, bool always_create = false);
 
+  static Status GetSharedModuleFromCAS(ConstString module_name,
```
Can you add a Doxygen comment for this?

adrian-prantl commented on lldb/include/lldb/Core/ModuleList.h:
```diff
@@ -522,6 +528,11 @@ class ModuleList {
                   bool *did_create_ptr, bool always_create = false);
 
+  static Status GetSharedModuleFromCAS(ConstString module_name,
+                                       const char *cas_id, FileSpec cu_path,
```
depending on where this comes from a StringRef/ConstString/std::string would be more idiomatic

adrian-prantl replied:
if it comes from DWARF that `const char *` is probably okay.

adrian-prantl commented on lldb/include/lldb/Core/ModuleList.h:
```diff
@@ -522,6 +528,11 @@ class ModuleList {
                   llvm::SmallVectorImpl<lldb::ModuleSP> *old_modules,
                   bool *did_create_ptr, bool always_create = false);
 
+  static Status GetSharedModuleFromCAS(ConstString module_name,
```
We try to move away from returning Status, and use llvm::Expected where possible. So maybe this should return Expected<ModuleSP>.

cachemeifyoucan replied:
`Status` is actually the best choice here. This is a parallel function for `GetSharedModule`, where it calls that function to finish the module load task, and also fallback to that function to load from file system if CAS loading failed.

For both dsymutil/lldb, we actually always tries to load CAS first, so it will be extremely noisy if we try to emit any error about CAS creation or CASID parsing, because those are all expected.

adrian-prantl replied:
I understand the CASID parsing part — that could be a special llvm::Error kind we ignore. But why would you want to suppress an error about CAS creation?

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -1277,6 +1310,45 @@ static SharedModuleList &GetSharedModuleList() {
+    shared_module_list.cas_object_store = std::move(maybe_cas->first);
+    return shared_module_list.cas_object_store;
+  } else
+    llvm::consumeError(maybe_cas.takeError());
```
We should probably bubble this error up so we can give a precise failure reason. Maybe joinError() it with a more descriptive higher-level error at a layer above.

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -1548,6 +1620,59 @@ ModuleList::GetSharedModule(const ModuleSpec &module_spec, ModuleSP &module_sp,
   return error;
 }
 
+static llvm::Error loadModuleFromCAS(ConstString module_name,
```
In other APIs in LLDB (FileSpec for example) we can this "Materialize" since it creates a local copy of the file.

cachemeifyoucan replied:
I added a new `SetData` to the module spec so I can just pass it through buffer, not writing to file at all.

adrian-prantl replied:
It's unfortunate we're changing the state of the object during its lifetime, but I agree that that is better than writing a temporary file.

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -1548,6 +1620,59 @@ ModuleList::GetSharedModule(const ModuleSpec &module_spec, ModuleSP &module_sp,
+        module_spec.GetFileSpec().GetPathAsConstString().GetStringRef());
+  });
+
+  return GetSharedModule(module_spec, module_sp, nullptr, nullptr, nullptr);
```
It would be nice if there were a variant that didn't need a temporary file and would just take a MemorBuffer. No idea how much work that is though.

adrian-prantl commented on lldb/test/Shell/SymbolFile/cas-gmodule.test:
```diff
@@ -0,0 +1,46 @@
+# RUN: %lldb %t/main -s %t/lldb.script 2>&1 | FileCheck %s
+# CHECK: loading module 'Bottom' using CASID
+# CHECK: loading module 'Top' using CASID
+# CHECK: top = (x = 1)
```
Like in the dsymutil patch, it would be important to also test at least the most common error paths (config file doesn't exist, build dir has been removed, CAS is using different file storage format,...)

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -913,6 +913,51 @@ static std::string GetClangModulesCacheProperty() {
+  cas_config.PluginPath =
+      ModuleList::GetGlobalModuleListProperties().GetCASPluginPath().GetPath();
+  cas_config.PluginOptions =
+      ModuleList::GetGlobalModuleListProperties().GetCASPluginOptions();
```
Can this be shared with the other use, and maybe move it to Module/ or Utils/

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1930,41 +1975,80 @@ void SwiftASTContext::AddExtraClangCC1Args(
-  // Ignore CAS info inside modules when loading.
-  invocation.getFrontendOpts().ModuleLoadIgnoreCAS = true;
-
   // Add options to allow clang importer to do implicit module build.
```
// Add options to allow ClangImporter to do implicit module build.

makes it easier to search

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1930,41 +1975,80 @@ void SwiftASTContext::AddExtraClangCC1Args(
-  invocation.getFrontendOpts().ModuleLoadIgnoreCAS = true;
-
   // Add options to allow clang importer to do implicit module build.
   invocation.getLangOpts().ImplicitModules = true;
```
why is this needed?

cachemeifyoucan replied:
comments added. This is just code moved, not newly added.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1930,41 +1975,80 @@ void SwiftASTContext::AddExtraClangCC1Args(
+    auto CheckModuleInCAS = [&](const std::string &key) {
+      auto id = m_cas->parseID(key);
+      if (!id) {
+        llvm::consumeError(id.takeError());
```
If there is no way to bubble this up passing it to SWIFT_HEALTH_LOG would be the next best thing. Or doing LLDB_LOG_ERROR(), which consumes the error, too.

adrian-prantl commented:
Conceptually this looks good, thanks!

cachemeifyoucan commented:
@adrian-prantl Can you take a look again?

Not all error conditions can be tested or checked due to the limitation of CAS interface. We will not know things like version mismatch. Also all CAS checks are done before checking file system, so we cannot throw messages when CAS lookup failed. I tried best to log/healthcheck most of the CAS loads so we can tell if the CAS read is successful or not with additional logging.

adrian-prantl commented on llvm/docs/CommandGuide/dsymutil.rst:
```diff
@@ -40,6 +40,21 @@ OPTIONS
 
+.. option:: --cas <path to CAS>
+
+ Specify the path to CAS where binaries can be loaded instead of file system.
```
Specify the path to a content addressable storage that dsymutil may use to resolve `llvmcas://` URL file paths in the debug info.

cachemeifyoucan replied:
CASID is not guaranteed to have `llvmcas://` if plugin is used.

adrian-prantl commented on llvm/tools/dsymutil/BinaryHolder.cpp:
```diff
@@ -144,6 +146,40 @@ Error BinaryHolder::ObjectEntry::load(IntrusiveRefCntPtr<vfs::FileSystem> VFS,
+  }
+
+  if (Opts.Verbose)
+    WithColor::note() << "try load cas object " << PossibleID << "\n";
```
WithColor::note() << "resolving cas object " << PossibleID << "\n";

adrian-prantl commented on lldb/include/lldb/Core/ModuleList.h:
```diff
@@ -100,6 +101,12 @@ class ModuleListProperties : public Properties {
   // END SWIFT
 
+  // START CAS
+  FileSpec GetCASOnDiskPath() const;
```
Doxygen comment?

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -1548,6 +1623,55 @@ ModuleList::GetSharedModule(const ModuleSpec &module_spec, ModuleSP &module_sp,
+
+  Log *log = GetLog(LLDBLog::Modules);
+  if (log != nullptr)
+    LLDB_LOGF(log, "loading module '%s' using CASID '%s'",
```
more modern:
```
LLDB_LOG(log, "loading module '{0}' using CASID '{1}'", module_name, cas_id)
```

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -1548,6 +1623,55 @@ ModuleList::GetSharedModule(const ModuleSpec &module_spec, ModuleSP &module_sp,
+  auto err = loadModuleFromCAS(module_name, cas_id, cu_path, module_spec);
+  if (err) {
+    auto error_str = toString(std::move(err));
+    LLDB_LOGF(GetLog(LLDBLog::Modules),
```
same here

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -913,6 +913,28 @@ static std::string GetClangModulesCacheProperty() {
+  auto maybe_cas = cas_config->createDatabases();
+  if (!maybe_cas) {
+    HEALTH_LOG_PRINTF("failed to create CAS: %s",
+                      toString(maybe_cas.takeError()).c_str());
```
We may want to even surface that as a `Debugger::ReportWarning() `so it shows up on the console?

cachemeifyoucan committed:
[lldb][SwiftASTContext] Teach clang module loading from CAS

Teach SwiftASTContext to load clang module dependencies from CAS instead
of FileSystem.

cachemeifyoucan commented:
Everything should be addressed, except a standalone dsymutil test case. I need to think how to write that.

cachemeifyoucan commented:
@swift-ci please test llvm

adrian-prantl reviewed (approved):
Thanks for adding the tests!
